File: src/main/java/com/thebluehats/server/game/managers/combat/DamageManager.java
package com.thebluehats.server.game.managers.combat;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.UUID;

import com.google.inject.Inject;
import com.thebluehats.server.core.modules.annotations.MirrorReference;
import com.thebluehats.server.game.enchants.Mirror;
import com.thebluehats.server.game.managers.enchants.CustomEnchantUtils;
import com.thebluehats.server.game.managers.enchants.EnchantProperty;
import com.thebluehats.server.game.managers.world.regionmanager.RegionManager;
import com.thebluehats.server.game.utils.DataInitializer;
import com.thebluehats.server.game.utils.EntityValidator;

import org.bukkit.Material;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.potion.PotionEffectType;

public class DamageManager implements EntityValidator, DataInitializer {
    private final HashMap<UUID, EventData> eventData = new HashMap<>();
    private final ArrayList<UUID> canceledEvents = new ArrayList<>();
    private final ArrayList<UUID> removeCriticalDamage = new ArrayList<>();
    private final EnchantProperty<Float> mirrorReflectionValues = new EnchantProperty<>(0f, .25f, .5f);

    private final Mirror mirror;
    private final CombatManager combatManager;
    private final CustomEnchantUtils customEnchantUtils;
    private final RegionManager regionManager;

    @Inject
    public DamageManager(@MirrorReference Mirror mirror, CombatManager combatManager,
            CustomEnchantUtils customEnchantUtils, RegionManager regionManager) {
        this.combatManager = combatManager;
        this.mirror = mirror;
        this.customEnchantUtils = customEnchantUtils;
        this.regionManager = regionManager;
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onHit(EntityDamageByEntityEvent event) {
        UUID damagerUuid = event.getDamager().getUniqueId();

        if (canceledEvents.contains(damagerUuid)) {
            event.setCancelled(true);
        } else {
            event.setDamage(getDamageFromEvent(event));
        }

        eventData.remove(damagerUuid);
        canceledEvents.remove(damagerUuid);
    }

    @Override
    public boolean validate(Entity... entities) {
        for (Entity entity : entities) {
            if (uuidIsInCanceledEvent(entity.getUniqueId())) {
                return false;
            }
        }

        return true;
    }

    @Override
    @EventHandler
    public void initializeDataOnPlayerJoin(PlayerJoinEvent event) {
        eventData.put(event.getPlayer().getUniqueId(), new EventData());
    }

    /*
     * ----------- Utility Methods ----------
     */
    public double getDamageFromEvent(EntityDamageByEntityEvent event) {
        return calculateDamage(event.getDamage(), event);
    }

    public double getFinalDamageFromEvent(EntityDamageByEntityEvent event) {
        return calculateDamage(event.getFinalDamage(), event);
    }

    /*
     * ---------- Manipulation methods --------------
     */
    public void addDamage(EntityDamageByEntityEvent event, double value, CalculationMode mode) {
        EventData eventValues = eventData.computeIfAbsent(event.getDamager().getUniqueId(), k -> new EventData());

        switch (mode) {
            case ADDITIVE:
                eventValues.setAdditiveDamage(eventValues.getAdditiveDamage() + value);
                break;
            case MULTIPLICATIVE:
                eventValues.setMultiplicativeDamage(eventValues.getMultiplicativeDamage() + value);
                break;
        }
    }

    public void reduceDamageByPercentage(EntityDamageByEntityEvent event, double value) {
        EventData data = eventData.computeIfAbsent(event.getDamager().getUniqueId(), k -> new EventData());

        if (data.getReductionAmount() == 1) {
            data.setReductionAmount(data.getReductionAmount() - (1 - value));

            return;
        }

        data.setReductionAmount(1 - data.getReductionAmount() * value);
    }

    public void doTrueDamage(Player target, double damage) {
        if (!customEnchantUtils.itemHasEnchant(mirror, target.getInventory().getLeggings())) {
            target.setHealth(Math.max(0, target.getHealth() - damage));
            target.damage(0);
        }
    }

    public void doTrueDamage(Player target, double damage, Player reflectTo) {
        int level = customEnchantUtils.getEnchantLevel(mirror, target.getInventory().getLeggings());

        combatManager.combatTag(target);

        if (!customEnchantUtils.itemHasEnchant(mirror, target.getInventory().getLeggings())) {
            if (target.getHealth() - damage < 0) {
                safeSetPlayerHealth(target, 0);
            } else {
                target.damage(0);

                safeSetPlayerHealth(target, target.getHealth() - damage);
            }
        } else if (level != 1) {
            if (reflectTo.getHealth() - (damage * mirrorReflectionValues.getValueAtLevel(level)) < 0) {
                safeSetPlayerHealth(target, 0);
            } else {
                reflectTo.damage(0);

                combatManager.combatTag(target);

                safeSetPlayerHealth(reflectTo,
                        Math.max(0, reflectTo.getHealth() - (damage * mirrorReflectionValues.getValueAtLevel(level))));
            }
        }
    }

    public void removeExtraCriticalDamage(EntityDamageByEntityEvent event) {
        removeCriticalDamage.add(event.getDamager().getUniqueId());
    }

    public void setEventAsCanceled(EntityDamageByEntityEvent event) {
        canceledEvents.add(event.getDamager().getUniqueId());
    }

    public boolean eventIsNotCancelled(EntityDamageByEntityEvent event) {
        return !canceledEvents.contains(event.getDamager().getUniqueId());
    }

    public boolean uuidIsInCanceledEvent(UUID uuid) {
        return canceledEvents.contains(uuid);
    }

    public boolean isCriticalHit(Player player) {
        return player.getFallDistance() > 0 && !((Entity) player).isOnGround()
                && player.getLocation().getBlock().getType() != Material.LADDER
                && player.getLocation().getBlock().getType() != Material.VINE
                && player.getLocation().getBlock().getType() != Material.WATER
                && player.getLocation().getBlock().getType() != Material.LAVA && player.getVehicle() == null
                && !player.hasPotionEffect(PotionEffectType.BLINDNESS);
    }

    private double calculateDamage(double initialDamage, EntityDamageByEntityEvent event) {
       EventData data = eventData.computeIfAbsent(event.getDamager().getUniqueId(), k -> new EventData());

        double damage = initialDamage * data.getAdditiveDamage() * data.getMultiplicativeDamage()
                * data.getReductionAmount();

        if (removeCriticalDamage.contains(event.getDamager().getUniqueId())) {
            damage *= (2D / 3D);
        }

        if (event.getEntity() instanceof Player) {
            Player damagee = (Player) event.getEntity();

            ItemStack leggings = damagee.getInventory().getLeggings();

            if (leggings != null) {
                List<String> pantsLore = leggings.getItemMeta().getLore();

                if (pantsLore != null) {
                    for (String line : pantsLore) {
                        if (line.contains("As strong as iron")) {
                            damage *= 0.88;
                        }
                    }
                }
            }
        }

        damage -= data.getHeartReductionAmount();

        if (damage <= 0)
            damage = 1;

        return damage;
    }

    private void safeSetPlayerHealth(Player player, double health) {
        if (!regionManager.entityIsInSpawn(player)) {
            if (health >= 0 && health <= player.getMaxHealth()) {
                player.setHealth(health);
            }
        }
    }

    public void addHeartDamageReduction(EntityDamageByEntityEvent event, int hearts) {
        EventData data = eventData.computeIfAbsent(event.getDamager().getUniqueId(), k -> new EventData());

        data.setHeartReductionAmount(data.getHeartReductionAmount() + hearts);
    }

    private final static class EventData {
        private double additiveDamage = 1;
        private double multiplicativeDamage = 1;
        private double reductionAmount = 1;
        private int heartReductionAmount;

        public int getHeartReductionAmount() {
            return heartReductionAmount;
        }

        public void setHeartReductionAmount(int heartReductionAmount) {
            this.heartReductionAmount = heartReductionAmount;
        }

        public double getAdditiveDamage() {
            return additiveDamage;
        }

        public void setAdditiveDamage(double additiveDamage) {
            this.additiveDamage = additiveDamage;
        }

        public double getMultiplicativeDamage() {
            return multiplicativeDamage;
        }

        public void setMultiplicativeDamage(double multiplicativeDamage) {
            this.multiplicativeDamage = multiplicativeDamage;
        }

        public double getReductionAmount() {
            return reductionAmount;
        }

        public void setReductionAmount(double reductionAmount) {
            this.reductionAmount = reductionAmount;
        }
    }
}

File: src/main/java/com/thebluehats/server/game/managers/combat/DamageManager.kt
package com.thebluehats.server.game.managers.combat

import com.google.inject.Inject
import com.thebluehats.server.core.modules.annotations.MirrorReference
import com.thebluehats.server.game.enchants.Mirror
import com.thebluehats.server.game.managers.enchants.CustomEnchantUtils
import com.thebluehats.server.game.managers.enchants.EnchantProperty
import com.thebluehats.server.game.managers.world.regionmanager.RegionManager
import com.thebluehats.server.game.utils.DataInitializer
import com.thebluehats.server.game.utils.EntityValidator
import org.bukkit.Material
import org.bukkit.entity.Entity
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.potion.PotionEffectType
import java.util.*

class DamageManager @Inject constructor(
    @param:MirrorReference private val mirror: Mirror, private val combatManager: CombatManager,
    private val customEnchantUtils: CustomEnchantUtils, private val regionManager: RegionManager
) : EntityValidator, DataInitializer {
    private val eventData = HashMap<UUID, EventData>()
    private val canceledEvents = ArrayList<UUID>()
    private val removeCriticalDamage = ArrayList<UUID>()
    private val mirrorReflectionValues = EnchantProperty(0f, .25f, .5f)

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onHit(event: EntityDamageByEntityEvent) {
        val damagerUuid = event.damager.uniqueId

        if (canceledEvents.contains(damagerUuid)) {
            event.isCancelled = true
        } else {
            event.damage = getDamageFromEvent(event)
        }

        eventData.remove(damagerUuid)
        canceledEvents.remove(damagerUuid)
    }

    override fun validate(vararg entities: Array<Entity>): Boolean {
        for (entity in entities) {
            if (uuidIsInCanceledEvent(entity.uniqueId)) {
                return false
            }
        }
        return true
    }

    @EventHandler
    override fun initializeDataOnPlayerJoin(event: PlayerJoinEvent) {
        eventData[event.player.uniqueId] = EventData()
    }

    /*
     * ----------- Utility Methods ----------
     */
    fun getDamageFromEvent(event: EntityDamageByEntityEvent): Double {
        return calculateDamage(event.damage, event)
    }

    fun getFinalDamageFromEvent(event: EntityDamageByEntityEvent): Double {
        return calculateDamage(event.finalDamage, event)
    }

    /*
     * ---------- Manipulation methods --------------
     */
    fun addDamage(event: EntityDamageByEntityEvent, value: Double, mode: CalculationMode?) {
        val eventValues = eventData.computeIfAbsent(event.damager.uniqueId) { k: UUID? -> EventData() }
        when (mode) {
            CalculationMode.ADDITIVE -> eventValues.additiveDamage = eventValues.additiveDamage + value
            CalculationMode.MULTIPLICATIVE -> eventValues.multiplicativeDamage =
                eventValues.multiplicativeDamage + value
        }
    }

    fun reduceDamageByPercentage(event: EntityDamageByEntityEvent, value: Double) {
        val data = eventData.computeIfAbsent(event.damager.uniqueId) { k: UUID? -> EventData() }
        if (data.reductionAmount == 1.0) {
            data.reductionAmount = data.reductionAmount - (1 - value)
            return
        }
        data.reductionAmount = 1 - data.reductionAmount * value
    }

    fun doTrueDamage(target: Player, damage: Double) {
        if (!customEnchantUtils.itemHasEnchant(mirror, target.inventory.leggings)) {
            target.health = Math.max(0.0, target.health - damage)
            target.damage(0.0)
        }
    }

    fun doTrueDamage(target: Player, damage: Double, reflectTo: Player) {
        val data = customEnchantUtils.getItemEnchantData(mirror, target.inventory.leggings)
        combatManager.combatTag(target)
        if (data.itemHasEnchant()) {
            val level = data.enchantLevel
            if (level > 1) {
                val finalDamage = damage * mirrorReflectionValues.getValueAtLevel(level)
                combatManager.combatTag(reflectTo)
                reflectTo.damage(0.0)
                safeSetPlayerHealth(reflectTo, Math.max(0.0, reflectTo.health - finalDamage))
            }
        } else {
            target.damage(0.0)
            safeSetPlayerHealth(target, Math.max(0.0, target.health - damage))
        }
    }

    fun removeExtraCriticalDamage(event: EntityDamageByEntityEvent) {
        removeCriticalDamage.add(event.damager.uniqueId)
    }

    fun setEventAsCanceled(event: EntityDamageByEntityEvent) {
        canceledEvents.add(event.damager.uniqueId)
    }

    fun eventIsNotCancelled(event: EntityDamageByEntityEvent): Boolean {
        return !canceledEvents.contains(event.damager.uniqueId)
    }

    fun uuidIsInCanceledEvent(uuid: UUID): Boolean {
        return canceledEvents.contains(uuid)
    }

    fun isCriticalHit(player: Player): Boolean {
        return (player.fallDistance > 0 && !(player as Entity).isOnGround
                && player.location.block.type != Material.LADDER && player.location.block.type != Material.VINE && player.location.block.type != Material.WATER && player.location.block.type != Material.LAVA && player.vehicle == null && !player.hasPotionEffect(
            PotionEffectType.BLINDNESS
        ))
    }

    private fun calculateDamage(initialDamage: Double, event: EntityDamageByEntityEvent): Double {
        val data = eventData.computeIfAbsent(event.damager.uniqueId) { k: UUID? -> EventData() }
        var damage = (initialDamage * data.additiveDamage * data.multiplicativeDamage
                * data.reductionAmount)
        if (removeCriticalDamage.contains(event.damager.uniqueId)) {
            damage *= 2.0 / 3.0
        }
        if (event.entity is Player) {
            val damagee = event.entity as Player
            val leggings = damagee.inventory.leggings
            if (leggings != null) {
                val pantsLore = leggings.itemMeta.lore
                if (pantsLore != null) {
                    for (line in pantsLore) {
                        if (line.contains("As strong as iron")) {
                            damage *= 0.88
                        }
                    }
                }
            }
        }
        damage -= data.heartReductionAmount.toDouble()
        if (damage <= 0) damage = 1.0
        return damage
    }

    private fun safeSetPlayerHealth(player: Player, health: Double) {
        if (!regionManager.entityIsInSpawn(player)) {
            if (health >= 0 && health <= player.maxHealth) {
                player.health = health
            }
        }
    }

    fun addHeartDamageReduction(event: EntityDamageByEntityEvent, hearts: Int) {
        val data = eventData.computeIfAbsent(event.damager.uniqueId) { k: UUID? -> EventData() }
        data.heartReductionAmount = data.heartReductionAmount + hearts
    }

    private class EventData {
        var additiveDamage = 1.0
        var multiplicativeDamage = 1.0
        var reductionAmount = 1.0
        var heartReductionAmount = 0
    }
}